File: packages/php/src/Resources/Numbers.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `numbers` resource: virtual numbers and their inbound routing rules. Everything except `list` and `available` needs a live key.
 * Accessed as `$client->numbers`.
 */
final class Numbers
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function list(array $params = []): Page
    {
        return Models::page($this->transport->request('GET', '/v1/numbers', Models::map($params, Models::PAGE_QUERY, [], true)));
    }

    /**
     * Numbers available to assign.
     *
     * @param array{country: string, kind: string} $params
     * @return list<array<string, mixed>>
     */
    public function available(array $params): array
    {
        $query = Models::map($params, Models::NUMBER_QUERY, [], true);

        return Models::list($this->transport->request('GET', '/v1/numbers/available', $query));
    }

    /**
     * Assign (rent) a number. Charges the wallet.
     *
     * @param array{country: string, kind: string} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function assign(array $params, array $options = []): array
    {
        $body = Models::map($params, Models::NUMBER_QUERY, ['country', 'kind']);

        return Models::object($this->transport->request('POST', '/v1/numbers', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    public function release(string $id): void
    {
        $this->transport->request('DELETE', '/v1/numbers/' . Models::pathParam($id));
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function listRules(string $id, array $params = []): Page
    {
        $query = Models::map($params, Models::PAGE_QUERY, [], true);

        return Models::page($this->transport->request('GET', '/v1/numbers/' . Models::pathParam($id) . '/rules', $query));
    }

    /**
     * Rule: `match` (keyword|prefix|regex|any), `pattern`, `action`
     * (webhook|auto_reply|forward_email), `target`, `position`.
     *
     * @param array<string, mixed> $rule
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function createRule(string $id, array $rule, array $options = []): array
    {
        $body = Models::map($rule, Models::NUMBER_RULE);

        return Models::object($this->transport->request('POST', '/v1/numbers/' . Models::pathParam($id) . '/rules', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * Replace a rule (PUT).
     *
     * @param array<string, mixed> $rule
     * @return array<string, mixed>
     */
    public function updateRule(string $id, string|int $ruleId, array $rule): array
    {
        $body = Models::map($rule, Models::NUMBER_RULE);
        $path = '/v1/numbers/' . Models::pathParam($id) . '/rules/' . Models::pathParam($ruleId, 'ruleId');

        return Models::object($this->transport->request('PUT', $path, [], $body));
    }

    public function deleteRule(string $id, string|int $ruleId): void
    {
        $this->transport->request('DELETE', '/v1/numbers/' . Models::pathParam($id) . '/rules/' . Models::pathParam($ruleId, 'ruleId'));
    }
}

File: packages/java/src/main/java/io/opensms/Numbers.java
package io.opensms;

import io.opensms.models.NumberRule;
import io.opensms.models.Page;
import io.opensms.models.VirtualNumber;

import java.util.List;
import java.util.Map;

/**
 * Virtual numbers and their inbound rules. Accessed as {@code client.numbers()}.
 * Everything except {@link #list} and {@link #available} needs a live key.
 */
public final class Numbers {

    private final ApiTransport t;

    Numbers(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of your numbers. */
    public Page<VirtualNumber> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<VirtualNumber> list(ListParams params) {
        return t.get("/v1/numbers", params.toQuery(), ApiTransport.pageOf(VirtualNumber.class));
    }

    /**
     * @param country ISO2 country.
     * @param kind    {@code long_code}, {@code short_code} or {@code toll_free}.
     * @return numbers available to assign.
     */
    public List<VirtualNumber> available(String country, String kind) {
        return t.get("/v1/numbers/available", Query.of().add("country", country).add("kind", kind),
                ApiTransport.listOf(VirtualNumber.class));
    }

    /**
     * Assign a number (charges the wallet).
     *
     * @param country ISO2 country.
     * @param kind    number kind.
     * @return the assigned number.
     */
    public VirtualNumber assign(String country, String kind) {
        return assign(country, kind, null);
    }

    /**
     * @param country ISO2 country.
     * @param kind    number kind.
     * @param options per-call options.
     * @return the assigned number.
     */
    public VirtualNumber assign(String country, String kind, RequestOptions options) {
        Map<String, Object> body = Wire.map();
        body.put("country", Wire.require(country, "country"));
        body.put("kind", Wire.require(kind, "kind"));
        return t.json("POST", "/v1/numbers", body, ApiTransport.idempotencyKey(options),
                ApiTransport.type(VirtualNumber.class));
    }

    /** @param id number id. */
    public void release(String id) {
        t.json("DELETE", "/v1/numbers/" + Wire.id(id, "id"), null, null, ApiTransport.VOID);
    }

    /**
     * @param id number id.
     * @return the first page of rules.
     */
    public Page<NumberRule> listRules(String id) {
        return listRules(id, new ListParams());
    }

    /**
     * @param id     number id.
     * @param params paging.
     * @return one page.
     */
    public Page<NumberRule> listRules(String id, ListParams params) {
        return t.get("/v1/numbers/" + Wire.id(id, "id") + "/rules", params.toQuery(),
                ApiTransport.pageOf(NumberRule.class));
    }

    /**
     * @param id   number id.
     * @param rule the rule.
     * @return the created rule.
     */
    public NumberRule createRule(String id, NumberRuleParams rule) {
        return createRule(id, rule, null);
    }

    /**
     * @param id      number id.
     * @param rule    the rule.
     * @param options per-call options.
     * @return the created rule.
     */
    public NumberRule createRule(String id, NumberRuleParams rule, RequestOptions options) {
        return t.json("POST", "/v1/numbers/" + Wire.id(id, "id") + "/rules", rule.toWire(),
                ApiTransport.idempotencyKey(options), ApiTransport.type(NumberRule.class));
    }

    /**
     * @param id     number id.
     * @param ruleId rule id.
     * @param rule   the replacement rule.
     * @return the updated rule.
     */
    public NumberRule updateRule(String id, String ruleId, NumberRuleParams rule) {
        return t.json("PUT", "/v1/numbers/" + Wire.id(id, "id") + "/rules/" + Wire.id(ruleId, "ruleId"),
                rule.toWire(), null, ApiTransport.type(NumberRule.class));
    }

    /**
     * @param id     number id.
     * @param ruleId rule id.
     */
    public void deleteRule(String id, String ruleId) {
        t.json("DELETE", "/v1/numbers/" + Wire.id(id, "id") + "/rules/" + Wire.id(ruleId, "ruleId"),
                null, null, ApiTransport.VOID);
    }
}
